### PR TITLE
Url reader

### PR DESCRIPTION
**Pull request by kneumiller**
Part of the pull requests for OPeNDAP:
'read_and_distribute_file_content' has been modified to check for a url in place of filenames. If the user requests data from a libdap server (given that the libdap libraries are installed) then the data will be pulled from the url server through libdap and converted to the proper format for use within aspect.

### For all pull requests:

* [ X ] I have followed the [instructions for indenting my code](../CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [ X ] I have tested my new feature locally to ensure it is correct.
* [ X ] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../tests/) directory.
* [ X ] I have added a changelog entry in the [doc/modules/changes](../doc/modules/changes) directory that will inform other users of my change.

### CODE REVIEW

File: source/utilities.cc [hunk 1]
```diff
@@ -21,6 +21,13 @@
 #include <aspect/utilities.h>
 #include <aspect/simulator_access.h>
 
+#ifdef HAVE_LIBDAP
+#include <D4Connect.h>
```
Comment (tjhei) [inline source/utilities.cc]: I am surprised that libdap does not put its header files into a folder. A header like Array.h has the potential for conflicts.

File: source/utilities.cc [hunk 2]
```diff
@@ -906,11 +913,14 @@ namespace aspect
...
     }
 
-
-    std::pair<double,double> real_spherical_harmonic( const unsigned int l,
-                                                      const unsigned int m,
-                                                      const double theta,
-                                                      const double phi)
+    //Evaluate the cosine and sine terms of a real spherical harmonic.
+    //This is a fully normalized harmonic, that is to say, inner products
+    //of these functions should integrate to a kronecker delta over
```
Comment (tjhei) [inline source/utilities.cc]: this change is unrelated to your PR (and we also only keep the comments in the header file). Can you undo the change, please?

File: source/utilities.cc [hunk 3]
```diff
@@ -938,12 +948,166 @@ namespace aspect
...
+      std::stringstream urlString;
 
       if (Utilities::MPI::this_mpi_process(comm) == 0)
         {
           // set file size to an invalid size (signaling an error if we can not read it)
           unsigned int filesize = numbers::invalid_unsigned_int;
 
+          /**
+           * Lookup and use data from the provided url
+           * Author: Kodi Neumiller
```
Comment (tjhei) [inline source/utilities.cc]: please remove this whole block, we typically don't put authorship into code blocks

File: source/utilities.cc [hunk 4]
```diff
@@ -938,12 +948,166 @@ namespace aspect
...
+           * Added: 10/9/18
+           **/
+          //----Only run if the user wishes to use the libdap packages----//
```
Comment (tjhei) [inline source/utilities.cc]: please remove this comment line

File: source/utilities.cc [hunk 5]
```diff
@@ -938,12 +948,166 @@ namespace aspect
...
+#ifdef HAVE_LIBDAP
+          //Check to see if the prm file will be reading data from the disk or
+          // from a provided URL
+          if (filename.find("http://") == 0 || filename.find("https://") == 0 || filename.find("file://") == 0)
+            {
+              libdap::Connect *url = 0;
+              url = new libdap::Connect(filename);
```
Comment (tjhei) [inline source/utilities.cc]: 1. Move declaration and assignment into a single line.
2. You never seem to delete this object again, meaning you are leaking memory. Why do you not do ``libdap::Connect url(filename);``?

File: source/utilities.cc [hunk 6]
```diff
@@ -938,12 +948,166 @@ namespace aspect
...
+              libdap::BaseTypeFactory factory;
+              libdap::DataDDS dds(&factory);
+              libdap::DAS das;
+
+              url->request_data(dds, "");
+              url->request_das(das);
+
+
+              //Array to store the url data
+              libdap::Array *urlArray;
```
Comment (tjhei) [inline source/utilities.cc]: can you move this declaration inside the loop where you use it? We typically declare variables in the narrowest scope possible.

File: source/utilities.cc [hunk 7]
```diff
@@ -938,12 +948,166 @@ namespace aspect
...
+                                               " Maybe it was not of the correct type?"));
+                    }
+                }
+
+
+              //Add the POINTS data that is required and found at the top of the data file.
+              // The POINTS values are set as attributes inside a table.
+              // Loop through the Attribute table to locate the points values within
+              std::vector<std::string> points;
+              libdap::AttrTable *table;
```
Comment (tjhei) [inline source/utilities.cc]: move declaration inside loop

File: source/utilities.cc [hunk 8]
```diff
@@ -938,12 +948,166 @@ namespace aspect
...
+              for (libdap::AttrTable::Attr_iter i = das.var_begin(); i != das.var_end(); i++)
+                {
+                  table = das.get_table(i);
+                  if (table->get_attr("points") != "")
+                    points.push_back(table->get_attr("points"));
+                  else
+                    break;
```
Comment (tjhei) [inline source/utilities.cc]: why break here?

File: source/utilities.cc [hunk 9]
```diff
@@ -938,12 +948,166 @@ namespace aspect
...
+                  urlString << "\n";
+                }
+
+              data_string = urlString.str();
+              filesize = data_string.size();
+            }
+
+          else
+            {
+              std::ifstream filestream(filename.c_str());
```
Comment (tjhei) [inline source/utilities.cc]: is this just a copy&paste from the block below? I think you can reorder your ``ifdef`` and your ``if (starts with http)`` to avoid this copy.

File: source/utilities.cc [hunk 10]
```diff
@@ -938,12 +948,166 @@ namespace aspect
                                      const MPI_Comm &comm)
     {
       std::string data_string;
+      std::stringstream urlString;
```
Comment (tjhei) [inline source/utilities.cc]: move this declaration into the scope and as close to the first place you use it.

File: doc/modules/changes/20200116_kneumiller_2 [hunk 11]
```diff
@@ -0,0 +1,6 @@
+New: The 'read_and_distribute_file_content' now checks if the requested file lookup is a url.
+If the user has the libdap libraries installed and is requesting data from a libdap server then the
+data is pulled from the server, converted to the proper format, and passed in.
+<br>
+(Kodi Neumiller, 2020/01/16)
+<br>
```
Comment (tjhei) [inline doc/modules/changes/20200116_kneumiller_2]: please remove the last line

File: source/utilities.cc (hunk 4 above)
Comment (kneumiller) [reply source/utilities.cc]: Did you want me to remove the comment line "//----Only run if the user wishes to use the libdap packages----//"?

File: source/utilities.cc (hunk 5 above)
Comment (kneumiller) [reply source/utilities.cc]: Fixed, thanks for pointing that out.

File: source/utilities.cc (hunk 8 above)
Comment (kneumiller) [reply source/utilities.cc]: I think that may have been leftover code. Removed it.

File: source/utilities.cc (hunk 9 above)
Comment (kneumiller) [reply source/utilities.cc]: I'm not sure if I can/know how. The reason it's copied is in case the user has libdap turned on but is trying to read a local file. It's currently setup so that the `if (starts with http)` is triggered in the #else (meaning libdap is turned off) then it throws an error. Is there a way to setup the ifdef so that it can read a local file if libdap is on or off?

File: source/utilities.cc (hunk 4 above)
Comment (tjhei) [reply source/utilities.cc]: yes

**Comment by tjhei** [comment]
You can restructure the code in the following way:
```
          if (filename.find("http://") == 0 || filename.find("https://") == 0 || filename.find("file://") == 0)
            {
#ifdef HAVE_LIBDAP
[your code to read using libdap]
#else
AssertThrow(false, ExcMessage(...))
#endif
  }
else
  {
    the normal code
  }
}
```

**Comment by kneumiller** [comment]
Oh man, thank you, can't believe I didn't see that.

Also I just realized, I may need to revert some of the declarations because many of them are used outside of the loop they're instantiated in.

**Comment by kneumiller** [comment]
Let me know if this looks okay. 

Thanks for the help!

File: doc/modules/changes/20200116_kneumiller_2 [hunk 12]
```diff
@@ -0,0 +1,5 @@
+New: The 'read_and_distribute_file_content' now checks if the requested file lookup is a url.
```
Comment (tjhei) [inline doc/modules/changes/20200116_kneumiller_2]:
```suggestion
New: The 'read_and_distribute_file_content' function now checks if the requested file lookup is a URL.
```

File: doc/modules/changes/20200116_kneumiller_2 [hunk 13]
```diff
@@ -0,0 +1,5 @@
...
+If the user has the libdap libraries installed and is requesting data from a libdap server then the
+data is pulled from the server, converted to the proper format, and passed in.
+<br>
+(Kodi Neumiller, 2020/01/16)
```
Comment (tjhei) [inline doc/modules/changes/20200116_kneumiller_2]: please hit return at the end of the last line.

**Comment by tjhei** [comment]
What files are available at test.opendap.org, will the currently used files stay around, and would one go about adding new ones?

**Comment by tjhei** [comment]
@gassmoeller can you take a look at this PR as well, please? How do you feel about the files under contrib/? Should we move them somewhere else? cookbooks/ maybe?

**Comment by kneumiller** [comment]
Thanks for all of the suggestions, I'll look them over and push up the changes.

File: contrib/opendap/OPeNDAP.md [hunk 14]
```diff
@@ -0,0 +1,157 @@
+
+# The OPeNDAP extension to ASPECT
+
+The OPeNDAP client extension to ASPECT has been tested on OSX (10.13,
```
Comment (bangerth) [inline contrib/opendap/OPeNDAP.md]: I think it would be useful if you started this section with a sentence or two what opendap actually is. People may read this file trying to find out.

File: source/utilities.cc [hunk 15]
```diff
@@ -944,38 +951,158 @@ namespace aspect
           // set file size to an invalid size (signaling an error if we can not read it)
           unsigned int filesize = numbers::invalid_unsigned_int;
 
-          std::ifstream filestream(filename.c_str());
-
-          if (!filestream)
+          //Check to see if the prm file will be reading data from the disk or
+          // from a provided URL
+          if (filename.find("http://") == 0 || filename.find("https://") == 0 || filename.find("file://") == 0)
```
[the remark this answers is not in the document]
Comment (kneumiller) [reply source/utilities.cc]: Yeah that sounds like a good alternative

**Comment by kneumiller** [comment]
> What files are available at test.opendap.org, will the currently used files stay around, and would one go about adding new ones?

We use test.opendap.org to store data used for testing. The data should remain there until it is manually removed. You would have to talk to James about adding new data, but it is a fairly easy process.

File: source/utilities.cc [hunk 16]
```diff
@@ -944,38 +951,158 @@ namespace aspect
...
+              urlString << "# POINTS:";
+              for (unsigned int i = 0; i < points.size(); i++)
+                {
+                  urlString << " " << points[i];
+                }
+              urlString << "\n";
+
+              //Add the values from the arrays into the stringstream. The values are passed in
+              // per row with a character return added at the end of each row.
+              // TODO: Add a check to make sure that each column is the same size before writing
```
[the remark this answers is not in the document]
Comment (kneumiller) [reply source/utilities.cc]: Still needs work. An error occurs when reading one of the columns.

File: source/utilities.cc [hunk 17]
```diff
@@ -944,34 +961,157 @@ namespace aspect
...
-              AssertThrow (false,
-                           ExcMessage (std::string("Reading of file ") + filename + " finished " +
-                                       "before the end of file was reached. Is the file corrupted or"
-                                       "too large for the input buffer?"));
-              return data_string; // never reached
+              MPI_Bcast(&filesize, 1, MPI_UNSIGNED, 0, comm);
+              AssertThrow(false,
+                          ExcMessage(std::string("Reading of file ") + filename + " failed. " +
+                                     "Make sure you have the dependencies for reading a url " +
+                                     "(when running cmake make sure -DLIBDAP_ON=ON)"));
```
Comment (tjhei) [inline source/utilities.cc]:
```suggestion
                                     "(run cmake with -D ASPECT_WITH_LIBDAP=ON)"));
```

File: source/utilities.cc [hunk 18]
```diff
@@ -944,34 +961,157 @@ namespace aspect
...
+
+              std::ifstream filestream(filename.c_str());
```
Comment (tjhei) [inline source/utilities.cc]: why are you opening a file here? Isn't this the case where it is a URL and we don't have libdap? I think you can delete this whole block of code.

Comment (kneumiller) [reply source/utilities.cc]: I think you're right, fixed that.

**Comment by tjhei** [comment]
thank you, @kneumiller . This is okay from my end. Would you mind squashing your commits into a single one?

**Comment by kneumiller** [comment]
I can squash more commits if needed. Thanks for all of the help, glad we could get this pushed up

**Comment by gassmoeller** [comment]
Looks good from my point of view except for the unanswered comment from Wolfgang about providing a sentence or two at the beginning of the Readme about what this extension actually does. Maybe start the readme with a sentence like the following:
```
This is the OPeNDAP client extension for ASPECT. OPeNDAP is ... (1 sentence). 
Using the OPeNDAP extension ASPECT can query data from remote servers and 
include it in local computations. ... (continue with the existing text)
```
Feel free to modify as appropriate.

**Comment by dsarahstamps** [comment]
This needs to say something about the EarthCube funded BALTO project. It’s the BALTO plugin, not OPeNDAP.

**Comment by tjhei** [comment]
Looks good, thank you everyone!
